### PR TITLE
[FIXED JENKINS-59786] Report all the health reports of the primary branch, not just the worst one

### PR DESCRIPTION
**Pull request by stephenc**
See https://issues.jenkins-ci.org/browse/JENKINS-59786

### CODE REVIEW

**Comment by kshultzCB** [comment]
@stephenc , thanks for asking. I've got a handful of questions. Background:

* I've [merged 168 into this PR](https://github.com/kshultzCB/branch-api-plugin/tree/169-with-168), built it, and installed it.
* I'm building a multibranch project based on [this repo](https://github.com/kshultzCB/creates-and-deletes). As you can see in the [master branch's commit history](https://github.com/kshultzCB/creates-and-deletes/commits/master), I've created some long running builds, some intentionally failing builds, and some successful ones. Build history for master looks like this now:

  image

* The repo I'm building also has a branch in it called [intentionally-broken-branch](https://github.com/kshultzCB/creates-and-deletes/tree/intentionally-broken-branch), which as its name implies, [fails its build](https://github.com/kshultzCB/creates-and-deletes/blob/intentionally-broken-branch/Jenkinsfile#L5).

OK, cool...but what I'm seeing here when I back out a level in Jenkins and hover over the MBP's weather icon, is this:
image

Which is telling me only about the failed builds in intentionally-broken-branch.

Is this the expected behavior?

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java [hunk 1]
```diff
@@ -98,8 +103,12 @@ public void given__multibranch_with_primary__when__reporting_health__then__non_e
...
             FreeStyleProject stable = prj.getItem("stable");
             assertThat("We now have the stable branch", stable, notNullValue());
-            assertThat(prj.getBuildHealthReports(), not(containsInAnyOrder(master.getBuildHealth())));
-            assertThat(prj.getBuildHealthReports(), containsInAnyOrder(stable.getBuildHealth()));
+            assertThat(prj.getBuildHealthReports(), not(containsInAnyOrder(
+                    master.getBuildHealthReports().stream().map(Matchers::is).collect(Collectors.toList())
+            )));
+            assertThat(prj.getBuildHealthReports(), containsInAnyOrder(
+                    stable.getBuildHealthReports().stream().map(Matchers::is).collect(Collectors.toList())
+            ));
```
Comment (kshultzCB) [inline src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: A `reason` in these two `assertThat`s would really help explain the condition being tested for.

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java [hunk 2]
```diff
@@ -77,7 +80,9 @@ public void given__multibranch_with_primary__when__reporting_health__then__prima
             prj.invalidateBuildHealthReports();
             FreeStyleProject master = prj.getItem("master");
             assertThat("We now have the master branch", master, notNullValue());
-            assertThat(prj.getBuildHealthReports(), containsInAnyOrder(master.getBuildHealth()));
+            assertThat(prj.getBuildHealthReports(), containsInAnyOrder(
+                    master.getBuildHealthReports().stream().map(Matchers::is).collect(Collectors.toList())
+            ));
```
Comment (kshultzCB) [inline src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: A `reason` in this `assertThat` would help explain the condition being tested for.

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java (hunk 1 above)
Comment (stephenc) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: The only reason for the ‘stream’,’map’,’collect’ is that the `containsInAnyOrder(List)` expects a list of matchers not a list of instances

Comment (stephenc) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: We want to verify the list has all the primary branches reports and none of the other’s

**Comment by bitwiseman** [comment]
@stephenc 
Did you answer Karl's question and I just didn't see it?

File: src/main/java/jenkins/branch/PrimaryBranchHealthMetric.java [hunk 3]
```diff
@@ -111,7 +116,23 @@ public void observe(Item item) {
...
                     reports = new ArrayList<>();
-                    reports.add(getHealthReport(item));
+                    reports.addAll(Util.fixNull(getHealthReports(item)));
+                }
+            }
+        }
+
+        public static List<HealthReport> getHealthReports(Item item) {
+            if (item instanceof Job) {
+                return ((Job<?,?>) item).getBuildHealthReports();
```
Comment (bitwiseman) [inline src/main/java/jenkins/branch/PrimaryBranchHealthMetric.java]: Are there performance effects or changing this from querying a single item to querying a set of items?

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java (hunk 2 above)
Comment (stephenc) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: we want to check that the folder's health reports are the health reports of the master branch. We do not care about the order of the health reports (though it should be the same as long as they all have different `%` values). We need to wrap each health report in a matcher to stop the generic types being wrong

File: src/main/java/jenkins/branch/PrimaryBranchHealthMetric.java (hunk 3 above)
Comment (stephenc) [reply src/main/java/jenkins/branch/PrimaryBranchHealthMetric.java]: Should be faster as the single report needs top get the list first

**Comment by stephenc** [comment]
@bitwiseman the original feature was added in https://github.com/jenkinsci/branch-api-plugin/pull/146

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java (hunk 2 above)
Comment (bitwiseman) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: @stephenc 
The request was that you update the `assertThat` call to include a `reason` string.

File: src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java (hunk 1 above)
Comment (bitwiseman) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: @stephenc  
Please add `reason` strings to the `assertThat` method calls in the code.

Comment (stephenc) [reply src/test/java/jenkins/branch/PrimaryBranchHealthMetricTest.java]: Personally I think the reason is redundant once you see past the noise of generic type erasure, but I have added reasons anyway

**Comment by stephenc** [comment]
@bitwiseman what's stopping the merge?

**Comment by bitwiseman** [comment]
Build failing?

**Pull request merged**